capstone/capapi/views.py

```diff
@@ -56,19 +61,118 @@ class CitationViewSet(viewsets.GenericViewSet, mixins.RetrieveModelMixin, mixins
 
 
 class CaseViewSet(viewsets.GenericViewSet, mixins.RetrieveModelMixin, mixins.ListModelMixin,):
-    """
-    Browse all cases
-    """
     serializer_class = serializers.CaseSerializer
     http_method_names = ['get']
+    lookup_field = 'slug'
     queryset = models.CaseMetadata.objects.all()
     renderer_classes = (renderers.BrowsableAPIRenderer, renderers.JSONRenderer)
-    lookup_field = 'slug'
+    permission_classes = (permissions.IsAPIUser,)
     filter_class = filters.CaseFilter
 
+    def download_many(self):
+        all_cases = self.queryset.all()
+        # get max number of cases to download from query
+        # set max number to daily max if unspecified
+        max_num = int(self.request.query_params.get('max', settings.API_CASE_DAILY_ALLOWANCE))
+
+        try:
+            to_filter = view_helpers.generate_filters_from_query_params(self.request.query_params)
+            cases = all_cases.filter(to_filter).order_by('decision_date')
+        except TypeError:
+            cases = all_cases.order_by('decision_date')
+
+        if cases.count() == 0:
+            return JsonResponse({
+                'message': 'Request did not return any results.'
+            }, status=404, )
+
+        # get page number for paginated requests
+        page_num = int(self.request.query_params.get('page', 1))
+
+        # pagination or max number
+        if max_num <= settings.API_DOWNLOAD_LIMIT:
+            cases = cases[:max_num]
+        else:
+            start_at = (page_num-1) * settings.API_DOWNLOAD_LIMIT
+            end_at = page_num * settings.API_DOWNLOAD_LIMIT
+            cases = cases[start_at:end_at]
+
+        # if request was already paginated
+        whitelisted_filters = view_helpers.get_whitelisted_case_filters()
+        blacklisted_case_count = len(set(cases) & set(all_cases.exclude(whitelisted_filters)))
```

Comment:
This isn't very efficient, right? If we call this endpoint without any filters, `set(all_cases.exclude(whitelisted_filters))` will build a set of all 5 million cases.

Here's my suggestion:

- Add a `whitelisted` boolean to the Jurisdiction model
- Use `all_cases = self.queryset.select_related('jurisdiction')` to make sure they have jurisdiction set without hitting the DB again.
- `blacklisted_case_count = len(c for c in case if not case.jurisdiction.whitelisted)`

Reply:
much better, thanks!